pkg/ruler/base/ruler.go

```diff
@@ -378,6 +377,12 @@ type sender interface {
 	Send(alerts ...*notifier.Alert)
 }
 
+func grafanaLinkForExpression(expr string) string {
+	escapedExpression := url.QueryEscape(strings.ReplaceAll(expr, `"`, `\"`))
+	str := `/explore?left={"queries":[{"editorMode":"code","expr":"%s","queryType":"range"}]}`
```

Comment:
Would love any insight on this

Reply:
@chaudum just to clarify, this is going to fail if:

- user never changed Explore data source and default data source is not this loki instance's data source 
- this loki instance's data source wasn't the last data source user used

So if the last data source that user used is this for loki instance, then it would still work and loki doesn't have to be the default data source. 

I don't know about any way how we could link to Loki data source without knowing it's uid. Moreover, users can have multiple loki data sources. So the only idea I have is to add a config to specify loki's data source uid and use that. 

But maybe @gelicia who works on Explore and Explore's url have better idea. 🙂

Reply:
@ivanahuckova thanks for the clarification. Indeed, that's what I concluded from my testing as well: if Loki was last selected, then this works fine, and if it fails, changing the datasource on the explore page keeps the query intact so that's the only extra step necessary to get the right result.

Adding a Loki uid would be a good solution, but do we want to go that far for this?

Reply:
> changing the datasource on the explore page keeps the query intact so that's the only extra step necessary to get the right result.

Just to clarify, this works only if the last data source was (different) loki or prometheus. If it is something else, it won't keep the query intact. 

Another idea I have is that if explore's url would include data source type, grafana could do a lookup and set correct data source type. We wouldn't guarantee that the selected loki data source is the correct one, but at least you'd have loki data source selected and could change it if it wasn't the right one. +cc @gelicia

Reply:
> Just to clarify, this works only if the last data source was (different) loki or prometheus. If it is something else, it won't keep the query intact.
My bad, thanks for making things clear :slightly_smiling_face: 

> Another idea I have is that if explore's url would include data source type, grafana could do a lookup and set correct data source type.
This sounds like a good enhancement that would be useful not only for this case.

Reply:
@chaudum @ivanahuckova Any updates on how we can move this PR forward?

Reply:
Hi @aminesnow, sorry for late reply. I talked to Logs and Explore team and we discussed different options. To get 100% correct link (with the correct data source), you need loki data source's uid. User already have to configure `ExternalURL`, so this would just be additional configuration. In case it wouldn't be set up, the data source part could be omitted, and the url would work only if:
- user never changed Explore data source and default data source is  this loki instance's data source
- this loki instance's data source was the last data source user used. But this is to be considered an hacky workaround and not something i would rely on.

I can create feature request for Explore team related to passing data source type and getting first data source uid for that data source type, which would improve experience if user didn't set loki data source uid, but we still think that right way to go is for user to have a chance to specify loki data source uid.

Reply:
Thanks for the detailed reply @ivanahuckova . 
So should I go ahead and add the data source UID in preparation for the feature, or should I leave it as is and come back later when a new PR once it's implemented?
If I were to add it here, what would the UID parameter in the URL look like?

Reply:
> So should I go ahead and add the data source UID in preparation for the feature, or should I leave it as is and come back later when a new PR once it's implemented?

I'd leave that for @chaudum or @dannykopping to decide. 

> If I were to add it here, what would the UID parameter in the URL look like?

So currently the query we are passing to url is:
```
{
  "editorMode":"code",
  "expr":"up",
  "queryType":"range"
 }
```

and with the loki data source uid, it would be: 
```
{
  "editorMode":"code",
  "expr":"up",
  "queryType":"range", 
  "datasource": {
    "type": "loki",
    "uid": "LokiDatasourceUID"
  }
```

Also I would consider removing `editorMode` which would mean that user-selected/preferred editorMode would be shown.

Reply:
I think adding the loki datasource UID as a config option sounds reasonable, but it should have a sensible default if undefined.

The base path relies on the `external_url` setting being set as well

```golang
f.Var(&cfg.ExternalURL, "ruler.external.url", "URL of alerts return path.")
```

And I think we'd need to update the description of this option to be more clear